src/App.js

```diff
@@ -1,21 +1,58 @@
-import React, { Component } from "react";
+import React from "react";
 import { Route, BrowserRouter, Switch } from "react-router-dom";
 import DashboardPage from "./pages/DashboardPage";
 import SignIn from "./modules/sign/components/Signin";
+import { createContext } from "react"
+import { useReducer} from "react"
+
+const reducer = (state, { type, payload }) => {
+  switch (type) {
+    case "LOGIN":
+      return { ...state, loading: true };
+
+    case "LOGIN_SUCESS":
+      return { ...state, user: payload, loading: false };
+
+    case "LOGIN_FAILED":
+      return { ...state, user: {}, error: payload, loading: false };
+
+    case "GITLAB_AUTH":
+      localStorage.setItem('auth', "gitlab")
```

Comment:
Reducers should only return new state data an should do nothing more. The localStorage set call should be done elsewhere